File: android/src/main/java/com/jacobbrasil/snapkit/SnapkitPlugin.java
package com.jacobbrasil.snapkit;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snap.corekit.utils.SnapUtils;
import com.snap.creativekit.SnapCreative;
import com.snap.creativekit.api.SnapCreativeKitApi;
import com.snap.creativekit.exceptions.SnapMediaSizeException;
import com.snap.creativekit.exceptions.SnapStickerSizeException;
import com.snap.creativekit.exceptions.SnapVideoLengthException;
import com.snap.creativekit.media.SnapMediaFactory;
import com.snap.creativekit.media.SnapPhotoFile;
import com.snap.creativekit.media.SnapSticker;
import com.snap.creativekit.media.SnapVideoFile;
import com.snap.creativekit.models.SnapContent;
import com.snap.creativekit.models.SnapLiveCameraContent;
import com.snap.creativekit.models.SnapPhotoContent;
import com.snap.creativekit.models.SnapVideoContent;
import com.snap.loginkit.BitmojiQuery;
import com.snap.loginkit.LoginStateCallback;
import com.snap.loginkit.SnapLogin;
import com.snap.loginkit.SnapLoginProvider;
import com.snap.loginkit.UserDataQuery;
import com.snap.loginkit.UserDataResultCallback;
import com.snap.loginkit.exceptions.LoginException;
import com.snap.loginkit.exceptions.UserDataException;
import com.snap.loginkit.models.BitmojiData;
import com.snap.loginkit.models.MeData;
import com.snap.loginkit.models.UserDataResult;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * SnapkitPlugin
 */
public class SnapkitPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    @Nullable
    private MethodChannel channel;

    @Nullable
    private Activity _activity;

    @NonNull
    private Activity requireActivity() {
        return Objects.requireNonNull(_activity, "Snapkit plugin is not attached to an activity.");
    }

    @Nullable
    private SnapCreativeKitApi snapCreativeKitApi;

    @NonNull
    private SnapCreativeKitApi getSnapCreativeKitApi() {
        if (snapCreativeKitApi == null) snapCreativeKitApi = SnapCreative.getApi(requireActivity());
        return snapCreativeKitApi;
    }

    @Nullable
    private SnapMediaFactory snapMediaFactory;

    @NonNull
    private SnapMediaFactory getSnapMediaFactory() {
        if (snapMediaFactory == null)
            snapMediaFactory = SnapCreative.getMediaFactory(requireActivity());
        return snapMediaFactory;
    }

    @Nullable
    private SnapLogin snapLogin;

    @NonNull
    private SnapLogin getSnapLogin() {
        if (snapLogin == null) snapLogin = SnapLoginProvider.get(requireActivity());
        return snapLogin;
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "snapkit");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull final Result result) {
        switch (call.method) {
            case "callLogin":
                callLogin(result);
                break;
            case "getUser":
                getUser(result);
                break;
            case "sendMedia":
                final String imagePath = call.argument("imagePath");
                final String videoPath = call.argument("videoPath");
                final Map<String, Object> sticker = call.argument("sticker");
                sendMedia(
                        Objects.requireNonNull(call.argument("mediaType")),
                        imagePath == null ? null : new File(imagePath),
                        videoPath == null ? null : new File(videoPath),
                        sticker == null ? null : StickerArguments.fromMap(sticker),
                        call.argument("caption"),
                        call.argument("attachmentUrl"),
                        result
                );
                break;
            case "verifyNumber":
                List<String> res = new ArrayList<>();
                res.add("");
                res.add("");
                result.success(res);
                break;
            case "callLogout":
                callLogout(result);
                break;
            case "isInstalled":
                result.success(SnapUtils.isSnapchatInstalled(
                        requireActivity().getPackageManager(),
                        "com.snapchat.android"
                ));
                break;
            case "getPlatformVersion":
                result.success("Android " + android.os.Build.VERSION.RELEASE);
                break;
            default:
                result.notImplemented();
                break;
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        Objects.requireNonNull(channel).setMethodCallHandler(null);
    }

    @Override
    public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
        _activity = binding.getActivity();
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {

    }

    @Override
    public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

    }

    @Override
    public void onDetachedFromActivity() {
        _activity = null;
    }

    // Creative Kit: https://docs.snap.com/snap-kit/creative-kit/Tutorials/android

    void sendMedia(
            @NonNull String mediaType,
            @Nullable File imagePath,
            @Nullable File videoPath,
            @Nullable StickerArguments sticker,
            @Nullable String caption,
            @Nullable String attachmentUrl,
            @NonNull MethodChannel.Result result
    ) {
        final SnapMediaFactory mediaFactory = getSnapMediaFactory();

        SnapContent content;
        switch (mediaType) {
            case "PHOTO":
                if (imagePath == null) {
                    result.error(
                            "SendMediaError",
                            "mediaType is set to photo but imagePath is null.",
                            null
                    );
                    return;
                }

                SnapPhotoFile photoFile;
                try {
                    photoFile = mediaFactory.getSnapPhotoFromFile(imagePath);
                } catch (SnapMediaSizeException e) {
                    result.error("SendMediaError", "Could not create SnapPhotoFile", e);
                    return;
                } catch (NullPointerException e) {
                    result.error("SendMediaError", "Could not find image file", e);
                    return;
                }

                content = new SnapPhotoContent(photoFile);
                break;
            case "VIDEO":
                if (videoPath == null) {
                    result.error(
                            "SendMediaError",
                            "mediaType is set to video but videoPath is null.",
                            null
                    );
                    return;
                }

                SnapVideoFile videoFile;
                try {
                    videoFile = mediaFactory.getSnapVideoFromFile(videoPath);
                } catch (SnapMediaSizeException | SnapVideoLengthException e) {
                    result.error("SendMediaError", "Could not create SnapVideoFile", e);
                    return;
                } catch (NullPointerException e) {
                    result.error("SendMediaError", "Could not find video file", e);
                    return;
                }

                content = new SnapVideoContent(videoFile);
                break;
            default:
                content = new SnapLiveCameraContent();
                break;
        }

        content.setCaptionText(caption);
        content.setAttachmentUrl(attachmentUrl);

        if (sticker != null) {
            SnapSticker snapSticker;
            try {
                snapSticker = mediaFactory.getSnapStickerFromFile(sticker.image);
            } catch (SnapStickerSizeException e) {
                result.error("SendMediaError", "Could not create SnapSticker", e);
                return;
            } catch (NullPointerException e) {
                result.error("SendMediaError", "Could not find sticker file", e);
                return;
            }

            snapSticker.setWidthDp(sticker.width);
            snapSticker.setHeightDp(sticker.height);
            snapSticker.setPosX(sticker.offsetX);
            snapSticker.setPosY(sticker.offsetY);
            snapSticker.setRotationDegreesClockwise(sticker.rotation);
            content.setSnapSticker(snapSticker);
        }

        getSnapCreativeKitApi().send(content);
    }

    static class StickerArguments {
        @NonNull
        File image;
        float width;
        float height;
        float offsetX;
        float offsetY;
        float rotation;

        public StickerArguments(
                @NonNull File image,
                float width,
                float height,
                float offsetX,
                float offsetY,
                float rotation
        ) {
            this.image = image;
            this.width = width;
            this.height = height;
            this.offsetX = offsetX;
            this.offsetY = offsetY;
            this.rotation = rotation;
        }

        static StickerArguments fromMap(Map<String, Object> map) {
            return new StickerArguments(
                    new File((String) Objects.requireNonNull(map.get("imagePath"))),
                    ((Double) Objects.requireNonNull(map.get("width"))).floatValue(),
                    ((Double) Objects.requireNonNull(map.get("height"))).floatValue(),
                    ((Double) Objects.requireNonNull(map.get("offsetX"))).floatValue(),
                    ((Double) Objects.requireNonNull(map.get("offsetY"))).floatValue(),
                    ((Double) Objects.requireNonNull(map.get("rotation"))).floatValue()
            );
        }
    }

    // Login Kit: https://docs.snap.com/snap-kit/login-kit/Tutorials/android

    void callLogin(@NonNull MethodChannel.Result result) {
        final SnapLogin snapLogin = getSnapLogin();
        final LoginStateCallback callback = new LoginStateCallback() {
            @Override
            public void onStart() {
            }

            @Override
            public void onSuccess(@NonNull String s) {
                result.success("Login Success");
                snapLogin.removeLoginStateCallback(this);
            }

            @Override
            public void onFailure(@NonNull LoginException e) {
                result.error("LoginError", "Error Logging In", e);
                snapLogin.removeLoginStateCallback(this);
            }

            @Override
            public void onLogout() {
                snapLogin.removeLoginStateCallback(this);
            }
        };
        snapLogin.addLoginStateCallback(callback);
        snapLogin.startTokenGrant();
    }

    void getUser(@NonNull MethodChannel.Result result) {
        final BitmojiQuery bitmojiQuery = BitmojiQuery.newBuilder()
                .withTwoDAvatarUrl()
                .build();
        final UserDataQuery userDataQuery = UserDataQuery.newBuilder()
                .withExternalId()
                .withDisplayName()
                .withBitmoji(bitmojiQuery)
                .build();
        getSnapLogin().fetchUserData(userDataQuery, new UserDataResultCallback() {
            @Override
            public void onSuccess(@NonNull UserDataResult userDataResult) {
                if (userDataResult.getData() == null)
                    return;


                final MeData meData = userDataResult.getData().getMeData();
                if (meData == null) {
                    result.error("GetUserError", "Returned MeData was null", null);
                    return;
                }
                final BitmojiData bitmojiData = meData.getBitmojiData();

                List<String> res = new ArrayList<>();
                res.add(meData.getExternalId());
                res.add(meData.getDisplayName());
                res.add(bitmojiData == null ? null : bitmojiData.getTwoDAvatarUrl());
                result.success(res);
            }

            @Override
            public void onFailure(@NonNull UserDataException e) {
                UserDataException.Status status = null;
                for (UserDataException.Status s : UserDataException.Status.values()) {
                    if (s.code == e.getStatusCode()) {
                        status = s;
                        break;
                    }
                }
                Objects.requireNonNull(status);
                result.error("UnknownGetUserError", status.message, status.code);
            }
        });
    }

    void callLogout(@NonNull MethodChannel.Result result) {
        final SnapLogin snapLogin = getSnapLogin();
        LoginStateCallback callback = new LoginStateCallback() {
            @Override
            public void onStart() {
            }

            @Override
            public void onSuccess(@NonNull String s) {
                snapLogin.removeLoginStateCallback(this);
            }

            @Override
            public void onFailure(@NonNull LoginException e) {
                result.error("LogoutError", "Error Logging In", e);
                snapLogin.removeLoginStateCallback(this);
            }

            @Override
            public void onLogout() {
                result.success("Logout Success");
                snapLogin.removeLoginStateCallback(this);
            }
        };
        snapLogin.addLoginStateCallback(callback);
        snapLogin.clearToken();
    }
}

File: android/src/main/kotlin/com/jacobbrasil/snapkit/SnapkitPlugin.kt
package com.jacobbrasil.snapkit

import android.app.Activity
import android.content.pm.PackageManager
import android.content.pm.PackageManager.PackageInfoFlags
import com.snap.creativekit.SnapCreative
import com.snap.creativekit.api.SnapCreativeKitApi
import com.snap.creativekit.api.SnapCreativeKitCompletionCallback
import com.snap.creativekit.api.SnapCreativeKitSendError
import com.snap.creativekit.exceptions.SnapMediaSizeException
import com.snap.creativekit.exceptions.SnapStickerSizeException
import com.snap.creativekit.exceptions.SnapVideoLengthException
import com.snap.creativekit.media.SnapMediaFactory
import com.snap.creativekit.media.SnapPhotoFile
import com.snap.creativekit.media.SnapSticker
import com.snap.creativekit.models.SnapContent
import com.snap.creativekit.models.SnapLiveCameraContent
import com.snap.creativekit.models.SnapPhotoContent
import com.snap.creativekit.models.SnapVideoContent
import com.snap.loginkit.AccessTokenResultCallback
import com.snap.loginkit.BitmojiQuery
import com.snap.loginkit.LoginResultCallback
import com.snap.loginkit.SnapLogin
import com.snap.loginkit.SnapLoginProvider
import com.snap.loginkit.UserDataQuery
import com.snap.loginkit.UserDataResultCallback
import com.snap.loginkit.exceptions.AccessTokenException
import com.snap.loginkit.exceptions.LoginException
import com.snap.loginkit.exceptions.UserDataException
import com.snap.loginkit.models.UserDataResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.util.Objects


/** SnapkitPlugin */
class SnapkitPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  private var _activity : Activity? = null

  private var _snapApi : SnapCreativeKitApi? = null
  private var _snapMediaFactory : SnapMediaFactory? = null

  private fun requireActivity(): Activity {
    return Objects.requireNonNull<Activity>(_activity, "Snapkit plugin is not attached to an activity.")
  }

  private fun getSnapApi(): SnapCreativeKitApi {
    if (_snapApi == null) {
      _snapApi = SnapCreative.getApi(requireActivity())
    }

    return _snapApi!!
  }

  private fun getSnapMediaFactory(): SnapMediaFactory {
    if (_snapMediaFactory == null) {
      _snapMediaFactory = SnapCreative.getMediaFactory(requireActivity())
    }

    return _snapMediaFactory!!
  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "snapkit")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "sdkVersion" -> {
        result.success(SnapLoginProvider.getVersion())
      }
      "isSnapchatInstalled" -> {
        try {
          val pm: PackageManager = requireActivity().packageManager
          pm.getPackageInfo("com.snapchat.android", PackageInfoFlags.of(0))
          result.success(true)
        } catch (e: PackageManager.NameNotFoundException) {
          result.success(false)
        }
      }
      "isLoggedIn" -> {
        result.success(SnapLoginProvider.get(requireActivity()).isUserLoggedIn)
      }
      "login" -> {
        SnapLoginProvider.get(requireActivity()).startTokenGrant(object: LoginResultCallback {
          override fun onStart() {
            // Nothing is done here
          }

          override fun onSuccess(accessToken: String) {
            result.success("Login Success")
          }

          override fun onFailure(exception: LoginException) {
            result.error("LoginError", exception.localizedMessage, null)
          }
        })
      }
      "getCurrentUser" -> {
        val bitmojiQuery = BitmojiQuery.newBuilder().withAvatarId().withTwoDAvatarUrl().build()
        val userDataQuery = UserDataQuery.newBuilder().withExternalId().withIdToken().withDisplayName().withBitmoji(bitmojiQuery).build()

        SnapLoginProvider.get(requireActivity()).fetchUserData(userDataQuery, object: UserDataResultCallback {
          override fun onSuccess(userDataResult: UserDataResult) {
            if (userDataResult.data?.meData == null) {
              result.error("GetUserError", "User data was null", null)
              return
            }

            val meData = userDataResult.data!!.meData!!
            val map: HashMap<String, String?> = HashMap<String, String?>()
            map["externalId"] = meData.externalId
            map["openIdToken"] = meData.idToken
            map["displayName"] = meData.displayName
            map["bitmoji2DAvatarUrl"] = meData.bitmojiData?.twoDAvatarUrl
            map["bitmojiAvatarId"] = meData.bitmojiData?.avatarId
            map["errors"] = null

            result.success(map)
          }

          override fun onFailure(exception: UserDataException) {
            result.error("GetUserError", exception.localizedMessage, exception)
          }
        })
      }
      "getAccessToken" -> {
        SnapLoginProvider.get(requireActivity()).fetchAccessToken(object: AccessTokenResultCallback {
          override fun onSuccess(token: String) {
            result.success(token)
          }

          override fun onFailure(e: AccessTokenException) {
            result.error("GetAccessTokenError", e.localizedMessage, e)
          }

        })
      }
      "logout" -> {
        SnapLoginProvider.get(requireActivity()).clearToken()
        result.success("Logout Success")
      }
      "shareToCamera" -> {
        if (call.arguments !is Map<*, *>) {
          return
        }

        val args = call.arguments as Map<*, *>

        try {
          val content = handleCommonShare(args, SnapLiveCameraContent())

          getSnapApi().sendWithCompletionHandler(content, object: SnapCreativeKitCompletionCallback {
            override fun onSendSuccess() {
              result.success("ShareToCamera Success")
            }

            override fun onSendFailed(e: SnapCreativeKitSendError?) {
              result.error("ShareToCameraError", e?.name, e)
            }
          })
        } catch (e: SnapStickerSizeException) {
          result.error("ShareToCameraError", e.localizedMessage, e)
        } catch (e: SnapKitException) {
          result.error("ShareToCameraError", e.localizedMessage, "Error caused by handleCommonShare")
        }
      }
      "shareWithPhoto" -> {
        if (call.arguments !is Map<*, *>) {
          return
        }

        val args = call.arguments as Map<*, *>

        try {
          val photo = File(args["photoPath"].toString())

          if (!photo.exists()) {
            result.error("ShareWithPhotoError","Photo could not be found in filesystem", null)
          }

          val content = handleCommonShare(args, SnapPhotoContent(getSnapMediaFactory().getSnapPhotoFromFile(photo)))

          getSnapApi().sendWithCompletionHandler(content, object: SnapCreativeKitCompletionCallback {
            override fun onSendSuccess() {
              result.success("ShareWithPhoto Success")
            }

            override fun onSendFailed(e: SnapCreativeKitSendError?) {
              result.error("ShareWithPhotoError", e?.name, e)
            }
          })
        } catch (e: SnapMediaSizeException) {
          result.error("ShareWithPhotoError", e.localizedMessage, e)
        } catch (e: SnapStickerSizeException) {
          result.error("ShareWithPhotoError", e.localizedMessage, e)
        } catch (e: SnapKitException) {
          result.error("ShareWithPhotoError", e.localizedMessage, "Error caused by handleCommonShare")
        }
      }
      "shareWithVideo" -> {
        if (call.arguments !is Map<*, *>) {
          return
        }

        val args = call.arguments as Map<*, *>

        try {
          val video = File(args["videoPath"].toString())

          if (!video.exists()) {
            result.error("ShareWithVideoError","Video could not be found in filesystem", null)
          }

          val content = handleCommonShare(args, SnapVideoContent(getSnapMediaFactory().getSnapVideoFromFile(video)))

          getSnapApi().sendWithCompletionHandler(content, object: SnapCreativeKitCompletionCallback {
            override fun onSendSuccess() {
              result.success("ShareWithVideo Success")
            }

            override fun onSendFailed(e: SnapCreativeKitSendError?) {
              result.error("ShareWithVideoError", e?.name, e)
            }
          })
        } catch (e: SnapMediaSizeException) {
          result.error("ShareWithVideoError", e.localizedMessage, e)
        } catch (e: SnapVideoLengthException) {
          result.error("ShareWithVideoError", e.localizedMessage, e)
        } catch (e: SnapStickerSizeException) {
          result.error("ShareWithVideoError", e.localizedMessage, e)
        } catch (e: SnapKitException) {
          result.error("ShareWithVideoError", e.localizedMessage, "Error caused by handleCommonShare")
        }
      }
      else -> {
        result.notImplemented()
      }
    }
  }

  private fun handleCommonShare(args: Map<*, *>, content: SnapContent): SnapContent {
    content.captionText = args["caption"].toString()
    content.attachmentUrl = args["link"].toString()

    if (args["sticker"] is Map<*, *>) {
      val sticker = args["sticker"] as Map<*, *>

      val image = File(sticker["imagePath"].toString())

      if (!image.exists()) {
        throw SnapKitException("Image could not be found in filesystem")
      }

      try {
        val snapSticker = getSnapMediaFactory().getSnapStickerFromFile(image)

        if (sticker["size"] is Map<*, *>) {
          val size = sticker["size"] as Map<*, *>
          snapSticker.setWidthDp(size["width"] as Float)
          snapSticker.setHeightDp(size["height"] as Float)
        }

        if (sticker["offset"] is Map<*, *>) {
          val offset = sticker["offset"] as Map<*, *>
          snapSticker.setPosX(offset["x"] as Float)
          snapSticker.setPosY(offset["y"] as Float)
        }

        if (sticker["rotation"] is Map<*, *>) {
          val rotation = sticker["rotation"] as Map<*, *>
          snapSticker.setRotationDegreesClockwise(rotation["angle"] as Float)
        }

        content.snapSticker = snapSticker
      } catch (e: SnapStickerSizeException) {
        throw e
      }
    }

    return content
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    _activity = binding.activity
  }

  override fun onDetachedFromActivity() {
    _activity = null
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {  }

  override fun onDetachedFromActivityForConfigChanges() {  }
}

class SnapKitException(message: String): Exception(message)
